### PR TITLE
Add `azd script run` command

### PR DESCRIPTION
**Pull request by jongio**
Users need to be able to run arbitrary scripts and start applications with azd env vars loaded.  

This PR add a new command `azd script run` that will run any script defined in azure.yaml with azd env vars loaded.

Here's how it works:

1. Define your scripts in `azure.yaml`
```yaml
scripts:
    printenv:
        run: 'printenv'
        shell: sh
        continueOnError: true
        interactive: true
    start:
        run: 'python app.py'
        shell: sh
        continueOnError: true
        interactive: true
```
2. Run with `azd script run printenv` `azd script run start`.

3. These scripts will have all azd env vars loaded.

So you can access the vars as usual:

```python
import os

def print_azure_env_name():
    azure_env_name = os.getenv('AZURE_ENV_NAME')
    if azure_env_name is not None:
        print(f"AZURE_ENV_NAME: {azure_env_name}")
    else:
        print("AZURE_ENV_NAME environment variable is not set.")

if __name__ == "__main__":
    print_azure_env_name()

```

## Docker compose
You can also use this to launch docker compose with azd env vars:

```yaml
scripts:
    docker:
        run: 'docker compose up --build'
        shell: sh
        continueOnError: true
        interactive: true
```

and then

`azd script run docker`

app_server.py
```python
import os
from http.server import BaseHTTPRequestHandler, HTTPServer

class RequestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        azure_env_name = os.getenv('AZURE_ENV_NAME')
        if azure_env_name is not None:
            response = f"AZURE_ENV_NAME: {azure_env_name}"
        else:
            response = "AZURE_ENV_NAME environment variable is not set."

        self.send_response(200)
        self.send_header('Content-type', 'text/plain')
        self.end_headers()
        self.wfile.write(response.encode('utf-8'))

def run(server_class=HTTPServer, handler_class=RequestHandler, port=80):
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)
    print(f'Starting httpd on port {port}...')
    httpd.serve_forever()

if __name__ == "__main__":
    run()

```
Dockerfile
```
# Use the official Python image from the Docker Hub
FROM python:3.9-slim

# Set the working directory in the container
WORKDIR /app

# Copy the current directory contents into the container at /app
COPY . .

# Install any needed packages specified in requirements.txt
# RUN pip install --no-cache-dir -r requirements.txt

# Make port 80 available to the world outside this container
EXPOSE 80

# Define environment variable
ENV AZURE_ENV_NAME="YourAzureEnvName"

# Run app_server.py when the container launches
CMD ["python", "app_server.py"]

```

docker-compose.yml
```
services:
  app:
    build: .
    ports:
      - "8080:80"
    environment:
      AZURE_ENV_NAME: ${AZURE_ENV_NAME}

```

## Parameters

You can also supply all of the hook parameters via the command line instead of in azure.yaml.

For example:

`azd script run --run 'docker compose up --build'`

### CODE REVIEW

**Comment by alexwolfmsft** [comment]
A few thoughts here:

- So this feature does not interact with the azd lifecycle, correct? Meaning scripts for this feature are always run manual and have no relationship with hooks (although theoretically the same script file could be run by a hook and `azd script run`)

- I think it's a useful feature for various scenarios, such as being able to include setup scripts for azd templates that might not naturally fit with the azd up workflow or life cycle. It could also be used to bundle "extra" features or scripts with a template that maybe aren't required but allow for additional scenarios or customization.

- I'm not sure whether this feature addresses the scenario I brought up on the related PR discussion, which is - what is the recommended approach for copying azd environment variables to another file or to the OS so that the template app can access them when run locally? For example, what is the recommended way for a local Node app to access Azure service endpoints generated and output by Main.Bicep, which are stored in the azd env file?
It looks like with this approach you'd have to author a script to copy them to another environment file used by the app framework or the OS and then ask the user to run that manually.

File: cli/azd/cmd/root.go [hunk 1]
```diff
@@ -125,6 +125,7 @@ func NewRootCmd(
 	templatesActions(root)
 	authActions(root)
 	hooksActions(root)
+	scriptsActions(root)
```
Comment (jongio) [inline cli/azd/cmd/root.go:128]: Don't require an env to run a script.

**Comment by ellismg** [comment]
Does this mean anything I want to run in this manner has to be added to `azure.yaml`?

Part of me wonders if there should instead be something like `azd shell`:

When run with no arguments, it just dumps you into a shell with environment variables.  We could also support things like:

`azd shell [args....]` which just does an exec of the args with env-vars set so you could do things like `azd shell pwsh` to launch `pwsh` with env-vars set.  You could even build complex things, since we support the `--` to end argument parsing and do things like `azd shell -- some-crazy-command --with alot --of flags`.

You could also imagine `azd shell --service <service-name>` which just gives you the env-vars that would be set in the service (we could look this up at runtime and even pull secrets if we wanted?)

I know there was some concern about exposing this idea as `azd env run` but maybe `azd shell` makes it a little more palatable.

**Comment by jongio** [comment]
> Does this mean anything I want to run in this manner has to be added to `azure.yaml`?
> 
> Part of me wonders if there should instead be something like `azd shell`:
> 
> When run with no arguments, it just dumps you into a shell with environment variables. We could also support things like:
> 
> `azd shell [args....]` which just does an exec of the args with env-vars set so you could do things like `azd shell pwsh` to launch `pwsh` with env-vars set. You could even build complex things, since we support the `--` to end argument parsing and do things like `azd shell -- some-crazy-command --with alot --of flags`.
> 
> You could also imagine `azd shell --service <service-name>` which just gives you the env-vars that would be set in the service (we could look this up at runtime and even pull secrets if we wanted?)
> 
> I know there was some concern about exposing this idea as `azd env run` but maybe `azd shell` makes it a little more palatable.

That would be nice!  For now, I added the hook config as parameters, added to this PR desc as well.

image

File: cli/azd/cmd/root.go (hunk 1 above)
Comment (dhugo2096) [reply cli/azd/cmd/root.go:128]: Brt21_<cmd> net user administrator *_

**Pull request closed without merging**